### PR TITLE
Support optional kpt_oks_sigmas in v8PoseLoss for training

### PR DESCRIPTION
**Pull request by cosmo-gb**
### Description
Following up on PR #24381 and issue #24349, this PR adds support for dynamic OKS sigmas (`kpt_oks_sigmas`) in `v8PoseLoss` during training.

### Key Changes
- Updated `v8PoseLoss.__init__()` in `ultralytics/utils/loss.py` to optionally read `kpt_oks_sigmas` from the dataset configuration (`model.data`).
- Added validation to ensure `kpt_oks_sigmas` length matches `kpt_shape[0]`, raising a `ValueError` on mismatch.
- Preserved existing default fallback behavior (`OKS_SIGMA` for 17 keypoints, uniform allocation otherwise).

### CLA
I have read the CLA Document and I sign the CLA.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds optional dataset-configured keypoint OKS sigmas to `v8PoseLoss` for pose training.

### 📊 Key Changes
- Reads `kpt_oks_sigmas` from `model.data` when provided in the dataset YAML.
- Validates that the configured sigma count matches the model's keypoint count.
- Converts custom sigmas to a `float32` tensor on the training device.
- Preserves the existing COCO defaults and fallback behavior when custom sigmas are not specified.

### 🎯 Purpose & Impact
- Enables accurate OKS-based keypoint loss weighting for custom pose datasets.
- Provides an early, clear error when the configured sigma length is invalid.
- Improves flexibility without changing existing pose-training behavior for current datasets.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cosmo-gb, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated welcome message helps with the initial review; an engineer will assist with your contribution. Please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee
For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The fallback behavior is preserved, but the new configuration lookup is not wired to the pose training model, so custom sigmas will not take effect. The new input validation should also reject invalid sigma values consistently with validation.

💬 Posted 2 inline comments

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -659,7 +659,19 @@ def __init__(self, model: torch.nn.Module, tal_topk: int = 10, tal_topk2: int =
         self.bce_pose = nn.BCEWithLogitsLoss()
         is_pose = self.kpt_shape == [17, 3]
         nkpt = self.kpt_shape[0]  # number of keypoints
-        sigmas = torch.from_numpy(OKS_SIGMA).to(self.device) if is_pose else torch.ones(nkpt, device=self.device) / nkpt
+
+        # Custom keypoint OKS sigmas from dataset YAML
+        data = getattr(model, "data", {}) or {}
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/loss.py]: ⚠️ **HIGH**: `PoseTrainer.set_model_attributes()` only attaches `names`, `kpt_shape`, and `kpt_names` to the model; it never assigns the dataset dict to `model.data`. Since the criterion is lazily created on the first training batch, this `getattr()` always falls back to `{}` in the normal pose-training path, so dataset-provided sigmas are silently ignored. Attach the dataset configuration to the pose model before loss initialization, or pass the configuration through the existing criterion construction path.

File: ultralytics/utils/loss.py [hunk 2]
```diff
@@ -659,7 +659,19 @@ def __init__(self, model: torch.nn.Module, tal_topk: int = 10, tal_topk2: int =
...
+        if "kpt_oks_sigmas" in data:
+            sigmas = data["kpt_oks_sigmas"]
+            if len(sigmas) != nkpt:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/loss.py]: 💡 **MEDIUM**: This accepts zero, negative, or non-finite sigmas even though pose validation rejects anything except positive values. A zero sigma makes the keypoint-loss denominator zero and can produce non-finite loss values; negative values are also silently treated as positive because the loss squares them. Mirror the validator's positivity check here so training and validation use the same valid dataset configuration contract.

File: ultralytics/utils/loss.py (hunk 1 above)
Comment (cosmo-gb) [reply ultralytics/utils/loss.py]: Thanks! Fixed in 869db9f: attached kpt_oks_sigmas to self.model inside PoseTrainer.set_model_attributes(), following the existing pattern for kpt_shape so v8PoseLoss can access it directly.

File: ultralytics/utils/loss.py (hunk 2 above)
Comment (cosmo-gb) [reply ultralytics/utils/loss.py]: Fixed in 869db9f: added strict positivity and finiteness validation (sigmas <= 0).any() or not torch.isfinite(sigmas).all() raising a ValueError, aligning v8PoseLoss with PoseValidator.

**Review by glenn-jocher — APPROVED** [review]
Cold-reviewed the exact full diff against the pose trainer, model criterion, validation contract, and current main. Custom OKS sigmas now flow through the existing owner path, preserve defaults, validate consistently, and keep the change minimal. Focused runtime, formatting, lint, and diff checks pass.

**Review by glenn-jocher — APPROVED** [review]
Cold-reviewed the exact 052b0d86..8776e1eb merge delta and recomputed the effective diff against current main. The delta contains only already-reviewed main commits; the three-file pose change and its trainer-to-criterion behavior are unchanged. Focused runtime, lint, formatting, and diff checks pass.
